### PR TITLE
fix when set null orgin will return null back

### PR DESCRIPTION
**Pull request by jiuker**
## Description

CROS must set  AllowedOrigins work for [*]

## How to test this PR?

`curl -v -XOPTIONS -H "Origin: null" https://host:9000/testing/README.md`
You will see the following header!
before:
`Access-Control-Allow-Origin: null`
after:
`Access-Control-Allow-Origin:  *`
## Types of changes
- [x] Bug fix ( #17409 )
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: internal/config/api/api.go [hunk 1]
```diff
@@ -281,3 +277,11 @@ func LookupConfig(kvs config.KVS) (cfg Config, err error) {
 
 	return cfg, nil
 }
+
+func GetCorsAllowOrigin(kvs config.KVS) []string {
+	corsAllowOrigin := strings.Split(env.Get(EnvAPICorsAllowOrigin, kvs.Get(apiCorsAllowOrigin)), ",")
```
Comment (shtripat) [inline internal/config/api/api.go]: Should we make the condition as `if len(corsAllowOrigin) == 0 || (len(corsAllowOrigin) == 1 && corsAllowOrigin[0] == "")` for better readability?

**Review by shtripat — APPROVED** [review]
lgtm

File: cmd/api-router.go [hunk 2]
```diff
@@ -549,5 +549,6 @@ func corsHandler(handler http.Handler) http.Handler {
 		AllowedHeaders:   commonS3Headers,
 		ExposedHeaders:   commonS3Headers,
 		AllowCredentials: true,
+		AllowedOrigins:   globalAPIConfig.getCorsAllowOrigins(),
```
Comment (harshavardhana) [inline cmd/api-router.go]: This is not correct, please revert.

File: internal/config/api/api.go [hunk 3]
```diff
@@ -194,7 +194,7 @@ func LookupConfig(kvs config.KVS) (cfg Config, err error) {
 	}
 
 	corsAllowOrigin := strings.Split(env.Get(EnvAPICorsAllowOrigin, kvs.Get(apiCorsAllowOrigin)), ",")
-	if len(corsAllowOrigin) == 0 {
+	if len(corsAllowOrigin) == 1 && corsAllowOrigin[0] == "" {
```
Comment (harshavardhana) [inline internal/config/api/api.go]:
```suggestion
	if len(corsAllowOrigin) == 1 && corsAllowOrigin[0] == "" {
```

This is not needed we should do it differently here.

```
corsList := env.Get(EnvAPICorsAllowOrigin, kvs.Get(apiCorsAllowOrigin))
if corsList == "" {
   corsAllowOrigin = []string{"*"} // defaults to '*'
} else {
   corsAllowOrigin = strings.Split(corsList, ",")
   for _, cors := range corsAllowOrigin {
       if cors == "" {
           return cfg, errors.New("invalid cors value")
       }
   }
} 
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by harshavardhana** [comment]
The reason why this is wrong

- AllowedOrigins []string: A list of origins a cross-domain request can be executed from. If the special * value is present in the list, all origins will be allowed. An origin may contain a wildcard (*) to replace 0 or more characters (i.e.: http://*.domain.com). The usage of wildcards implies a small performance penalty. Only one wildcard can be used per origin. The default value is *.

- AllowOriginFunc func (origin string) bool: A custom function to validate the origin. It takes the origin as an argument and returns true if allowed, or false otherwise. If this option is set, the content of AllowedOrigins is ignored.

If you look at the `rs/cors` library setting AllowedOrigins will break AllowOriginFunc basically breaking existing users - you need a different change such as this to make this fully complete

```diff
diff --git a/cmd/api-router.go b/cmd/api-router.go
index 517b0dd6d..875c572f5 100644
--- a/cmd/api-router.go
+++ b/cmd/api-router.go
@@ -524,13 +524,9 @@ func corsHandler(handler http.Handler) http.Handler {
                "*",
        }
 
-       return cors.New(cors.Options{
+       opts := cors.Options{
                AllowOriginFunc: func(origin string) bool {
-                       allowedOrigins := globalAPIConfig.getCorsAllowOrigins()
-                       if len(allowedOrigins) == 0 {
-                               allowedOrigins = []string{"*"}
-                       }
-                       for _, allowedOrigin := range allowedOrigins {
+                       for _, allowedOrigin := range globalAPIConfig.getCorsAllowOrigins() {
                                if wildcard.MatchSimple(allowedOrigin, origin) {
                                        return true
                                }
@@ -549,5 +545,15 @@ func corsHandler(handler http.Handler) http.Handler {
                AllowedHeaders:   commonS3Headers,
                ExposedHeaders:   commonS3Headers,
                AllowCredentials: true,
-       }).Handler(handler)
+       }
+
+       for _, origin := range globalAPIConfig.getCorsAllowOrigins() {
+               if origin == "*" {
+                       opts.AllowOriginFunc = nil
+                       opts.AllowedOrigins = globalAPIConfig.getCorsAllowOrigins()
+                       break
+               }
+       }
+
+       return cors.New().Handler(handler)
 }
```

**Comment by harshavardhana** [comment]
However I feel the `rs/cors` library is doing a bit of magic here and is not documenting itself correctly on what is to be expected.

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**